src/diric.jl

```diff
@@ -0,0 +1,47 @@
+export diric
+
+"""
+    kernel = diric(Ω::Real, n::Int)
+
+Dirichlet kernel, also known as periodic sinc function,
+where `n` should be a positive integer.
+Returns `sin(Ω * n/2) / (n * sin(Ω / 2))` typically,
+but `±1` when `Ω` is a multiple of 2π.
+
+In the usual case where 'n' is odd, the output is equivalent to
+`1/n \\sum_{k=-(n+1)/2}^{(n+1)/2} e^{i k Ω}`
+which is the same as the discrete-time Fourier transform (DTFT)
+of a `n`-point moving average filter.
+
+When `n` is odd or even, the function is 2π or 4π periodic, respectively.
+
+When `Ω` is an `AbstractFloat` (e.g., `Float32` or `BigFloat`),
+the return type matches that of `Ω`.  Otherwise the return type is `Float64`.
+
+As of 2021-03-19, the Wikipedia definition has different factors.
+The definition here is consistent with scipy and other software frameworks.
+This implementation treats inputs near multiples of 2π fairly carefully.
+"""
+function diric(Ω::T, n::Int) where T <: AbstractFloat
+    n <= 0 && throw(ArgumentError("n=$n is non-positive"))
+    denom = sin(Ω / 2)
+    if isapprox(1 + denom, 1) # denom ≈ 0 ?
```

Comment:
This is a rather indirect condition. I'd propose `isapprox(denom, 0; atol=...)` instead, where I'm not sure what to use for the `...` here to be equivalent. (Which is part of the problem I'm having with this as it is now). Actually, there is no value that is exactly equivalent as
```julia
julia> denom = 1.4901161342859267e-8
1.4901161342859267e-8

julia> isapprox(1 + denom, 1)
true

julia> denom = -denom
-1.4901161342859267e-8

julia> isapprox(1 + denom, 1)
false
```
Whereas the proposed alternative would be symmetric around zero (which also seems to make more sense.)

Reply:
Very good points.  I agree that `atol` is preferable here for symmetry and clarity.
Because `isapprox` has a lot of unnecessary stuff in it here, the updated PR uses an `atol` that is provided as a keyword option with the most sensible default (that stems from `isapprox` code).
It is pretty unlikely that any user will ever use anything other than that default,
but I prefer not to bury constants inside functions...  See what you think about the new version.
Thanks!

Reply:
Actually, letting this sink in for a bit, I don't think `atol` should be a parameter. The function should automatically pick a value based on the input type. This is not a parameter one might want to change based on context. Ideally, it should be chosen at the cross-over point where the naive formula becomes less exact than the Taylor approximation.

So I tried to figure out where that is. Around zero, it seems the error of the naive formula should be bounded by `2eps(T)`. (And some numerical experiments seem to verify this.) Given that the correct result is close to one, just the numerical error of evaluating the Taylor approximation might be comparable, so even in a range where the error due to the truncation of the series is small, the Taylor approximation might still not be beneficial. I'm not sure what happens when subnormals come into play, though. But in that case, probably `sin(x)==x`, so the naive formula would give `1.0` which would also probably the closest floating point value. So that indicates that only the case where `Ω/2==0` would require special treatment. A small but non-zero `atol` would feel safer though; `atol=eps(T)` comes to mind. Or we estimate the error of the series truncation by the first omitted term and see where that exceeds `eps(T)`, but that might be overly complicated.

For odd `n`, if we first map Ω to within (-π,π) (as already suggested below), the behavior around zero would all we have to think about. For even `n`, there is also 2π to consider...

Reply:
Great points!  I agree that any decisions about tolerances should be in the design of the function rather than leaving it to the caller.